text/0000-type-parameter-lifetimes.md

```diff
@@ -0,0 +1,175 @@
+- Feature Name: Type parameter lifetimes
+- Start Date: 2023-07-11
+- RFC PR: [rust-lang/rfcs#3457](https://github.com/rust-lang/rfcs/pull/3457)
+- Rust Issue: [rust-lang/rust#3457](https://github.com/rust-lang/rust/issues/3457)
+
+# Summary
+[summary]: #summary
+
+Add `'Self` and `'T` lifetimes for every type parameter `T`.
+
+# Motivation
+[motivation]: #motivation
+
+Rust types may embed lifetimes, for example:
+```rust
+trait Greet {
+    fn greet(&self);
+}
+
+struct Person<'a> {
+    name: &'a str,
+}
+
+impl<'a> Greet for Person<'a> {
+    fn greet(&self) {
+        println!("Hello, {}!", self.name)
+    }
+}
+```
+
+Where that lifetime is known we can already accomodate this:
+```rust
+impl<'a> Person<'a> {
+    // Return value implicitly has lifetime 'a
+    fn boxed(self) -> Box<Person<'a>> {
+        Box::new(self)
+    }
+
+    // Return value explicitly has lifetime 'a
+    fn as_boxed_greeter(self) -> Box<dyn Greet + 'a> {
+        Box::new(self)
+    }
+}
+```
+
+In other cases we currently cannot. (According to [RFC 599], `Box<dyn Greet>` resolves to `Box<dyn Greet +'static>` in these contexts.)
+```rust
+trait BadBoxer: Greet {
+    // Valid declaration, but not implementable for Person<'a>:
+    fn as_boxed(self) -> Box<dyn Greet>;
+}
+
+// Valid function, not usable for Person<'a>:
+fn box_static_greeter<G: Greet + 'static>(g: G) -> Box<dyn Greet> {
+    Box::new(g)
+}
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In a trait declaration or implementation, `'Self` refers to the lifetime of the implementing type. We may use this as follows.
+```rust
+trait Greet {
+    fn greet(&self);
+
+    fn as_boxed(self) -> Box<dyn Greet + 'Self>;
+}
+
+struct Person<'a> {
+    name: &'a str,
+}
+
+impl<'a> Greet for Person<'a> {
+    fn greet(&self) {
+        println!("Hello, {}!", self.name)
+    }
+
+    // Here 'Self = 'a
+    fn as_boxed(self) -> Box<dyn Greet + 'Self> {
+       Box::new(self)
+    }
+}
+```
+
+For every type parameter `T`, `'T` refers to the lifetime of the type parameter.
+```rust
+// Here 'G is the lifetime of the input type:
+fn box_greeter<G: Greet>(g: G) -> Box<dyn Greet + 'G> {
+    Box::new(g)
+}
+
+#[test]
+fn test_box_greeter() {
+    fn foo(name: &str) -> Box<dyn Greet> {
```

Comment:
this has a mismatched return type. `foo`'s body returns `()` since the last statement ends with a `;`